### DISCUSSION TITLE
Connection are not released when a large number of engine.connect()

### DISCUSSION THREAD

**Discussion by qwertyue** [sqlalchemy/sqlalchemy#8672] [Usage Questions]
### Describe the bug

i have a frequently invoked request that does creates a child thread to do the following every time:
```
conn = engine.connect()

try:
    my_method()
except:
// catch error
finally:
    conn.close()
```
```
def new_method():
    conn = engine.connect()

    try:
        my_method()
    except:
    // catch error
    finally:
        conn.close()
```

Simply put, engine.connect() is called twice in a sub-thread.
However, if a large number of request arrive, that is, the check-out of the connect reaches the limit ( size + overflow) and the conenction is not released.

At this point, subsequent subthreads are blocked at the engine.connect() method.

### To Reproduce

The parameters used for creat_engine are as follows:
```python
create_engine(SQL, pool_size=5, max_overflow=100,
                                    pool_recycle=1, pool_timeout=60 * 60 * 24, poolclass=QueuePool  , echo_pool='debug', pool_pre_ping=True).execution_options(autocommit=True)
````

```python
def mainTest():
    ps = []
    for i in range(5):
        ps.append(multiprocessing.Process(target=subTest))
    for p in ps:
        p.start()
    for p in ps:
        p.join()

def subTest():
    ts = []
    for i in range(10):
        ts.append(threading.Thread(target=callRequest))
    for t in ts:
        t.start()
    for t in ts:
        t.join()

def callRequest:
    while True:
        // to call request
```

First, create five subthreads, and then create 10 threads in each subthread to send requests.

### Error

There's no error, but there's a lot of checkout connections, like this:

INFO: 	size = 5, checkedin = 0, overflow = 100, checkedout = 105
INFO: 	current thread number：1144

### Versions

- OS: Win10 21H2
- Python: 3.10.7
- SQLAlchemy: 1.4.42
- Database: Oracle 11
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): cx_oracle

**Comment by CaselIT** [comment]
Hi,

I'm not sure what's the question or issue here.
you mention:
```
INFO: size = 5, checkedin = 0, overflow = 100, checkedout = 105
INFO: current thread number：1144
```
if you have 1144 threads and 105 possible connection then that's obvious that not all thread can have a connection.
either reduce the number of thread or increase the possible connections

**Reply by qwertyue** [reply]
> Hi,
> 
> I'm not sure what's the question or issue here. you mention:
> 
> ```
> INFO: size = 5, checkedin = 0, overflow = 100, checkedout = 105
> INFO: current thread number：1144
> ```
> 
> if you have 1144 threads and 105 possible connection then that's obvious that not all thread can have a connection. either reduce the number of thread or increase the possible connections

The thread keeps waiting for the connection to be released. (Actually, 105 connections are not released.)

If only one connection created within one thread, then the connection will be released normally.

**Reply by qwertyue** [reply]
I got the cause of the problem, because the wrong sequence of resource requests resulted in deadlocks.

**Reply by CaselIT** [reply]
You have not provided the code that is runned by the thread, so it's hard to guess what may be causing it.

In general you could deadlock without coordination if the thread use two connections since if you start 105 threads they will all get a connection, exhaust the pool then all wait to get the second connection but the pool is full.

Either reduce the number of threads, increase the possible connection, coordinate the threads with something like a semaphore or use a single connection in the thread
